resources/aws_ec2.go

```diff
@@ -1014,6 +994,91 @@ func (obj *AwsEc2Res) prependName() string {
 	return AwsPrefix + obj.GetName()
 }
 
+// waitForStopAndTerminate waits for the instance to stop and waits for it to
+// terminate. If either waiter returns, the instance state is checked, and an
+// awsEc2Event is returned.
+func (obj *AwsEc2Res) waitForStopAndTerminate(c *ec2.EC2, ctx context.Context, waitInput *ec2.DescribeInstancesInput) (awsEc2Event, error) {
+	if err := waitUntilInstanceStoppedOrTerminatedWithContext(c, ctx, waitInput); err != nil {
+		return awsEc2EventNone, errwrap.Wrapf(err, "error waiting for instance to stop or terminate")
+	}
+	// Check the instance state, and return the appropriate event.
+	stateOutput, err := obj.client.DescribeInstances(waitInput)
+	if err != nil {
+		return awsEc2EventNone, errwrap.Wrapf(err, "error describing instances")
+	}
+	if len(stateOutput.Reservations) == 1 {
+		switch *stateOutput.Reservations[0].Instances[0].State.Name {
+		case "stopped":
+			return awsEc2EventInstanceStopped, nil
+		case "terminated":
+			return awsEc2EventInstanceTerminated, nil
+		}
+	}
+	return awsEc2EventNone, nil
+}
+
+// waitUntilInstanceStoppedOrTerminatedWithContext combines the two waiters
+// required to trigger events if a running instance is stopped or terminated.
+// This function is needed, because the AWS api only provides waiters for one
+// or the other.
+func waitUntilInstanceStoppedOrTerminatedWithContext(c *ec2.EC2, ctx context.Context, waitInput *ec2.DescribeInstancesInput) error {
+	errChan := make(chan error)
+	defer close(errChan)
+	wg := sync.WaitGroup{}
+	defer wg.Wait()
+	closeChan := make(chan struct{})
+	wgCtx, wgCancel := context.WithCancel(ctx)
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		select {
+		case <-wgCtx.Done():
```

Comment:
@purpleidea Not %100 sure if this is the right approach to closing closeChan. Is there a chance that this could block? IOW, if there's an error, is it possible that wgCtx won't get cancelled so closeChan doesn't get closed?